caffe2/python/model_helper.py

```diff
@@ -143,6 +143,7 @@ def __init__(self, name=None, init_params=True, allow_not_known_ops=True,
             'use_cudnn': True,
             'cudnn_exhaustive_search': False,
         }
+        self.param_to_float_copy = {}
```

Comment:
this doesn't seem to be very generic to put into the main modeling class and then having all the optimizers and modeling helpers access it... You also have quite a bit of copy paste in each of the helpers (fc and conv in this case)
Seems like you just need to have some metadata on parameters. There is a ParamterInfo class, but we are rethinking it right now. Parameters can have a bunch of things associated with them. Including optimizer specific information which seems you need here.


Another approach I thought of is that if BlobReference has some info  of  anticipated type of the blob. Then you could even automatically insert conversions to and from fp16 when operations like MomentumSGD are called. 

I would like to chat with Andrey Malevich from our team who is working on new parameter abstractions for ModelHelper class and see if this kind of info fits into his new stuff.